File: Plugins/VideoPlayer/src/ios/VideoPlayer/VideoAnalytics.swift
//
//  VideoAnalytics.swift
//  AnalyticsProto
//
//  Created by Gary Griswold on 6/6/17.
//  Copyright © 2017 ShortSands. All rights reserved.
//

import Foundation
import CoreMedia
import AWS

class VideoAnalytics {
    
    static let APP_NAME = "ShortSandsBible"
    
    var dictionary = [String: String]()
    let dateFormatter = ISO8601DateFormatter()
    
    let mediaSource: String
    let mediaId: String
    let languageId: String
    let silLang: String
    let sessionId: String
    
    // Pass following from play to playEnd
    var timeStarted: Date
    var mediaViewStartingPosition: Float64
    
     init(mediaSource: String,
          mediaId: String,
          languageId: String,
          silLang: String) {
        
        self.mediaSource = mediaSource
        self.mediaId = mediaId
        self.languageId = languageId
        self.silLang = silLang
        
        let analyticsSessionId = AnalyticsSessionId()
        self.sessionId = analyticsSessionId.getSessionId()
        
        self.timeStarted = Date()
        self.mediaViewStartingPosition = 0.0
    }
    
    func playStarted(position: CMTime) -> Void {
        self.dictionary.removeAll()
        self.dictionary["sessionId"] = self.sessionId
        self.dictionary["mediaSource"] = self.mediaSource
        self.dictionary["mediaId"] = self.mediaId
        self.dictionary["languageId"] = self.languageId
        self.dictionary["silLang"] = self.silLang
        
        let locale = Locale.current
        self.dictionary["language"] = locale.languageCode
        self.dictionary["country"] = locale.regionCode
        self.dictionary["locale"] = locale.identifier
        
        let device = UIDevice.current
        self.dictionary["deviceType"] = "mobile"
        self.dictionary["deviceFamily"] = "Apple"
        self.dictionary["deviceName"] = device.model
        self.dictionary["deviceOS"] = "ios"
        print("system name \(device.systemName)")
        self.dictionary["osVersion"] = device.systemVersion
        self.dictionary["appName"] = VideoAnalytics.APP_NAME
        
        let bundle = Bundle.main
        let info = bundle.infoDictionary
        self.dictionary["appVersion"] = info?["CFBundleShortVersionString"] as? String

        self.timeStarted = Date()
        self.dictionary["timeStarted"] = dateFormatter.string(from: self.timeStarted)

        self.dictionary["isStreaming"] = "true" // should this be 1 instead
        self.mediaViewStartingPosition = CMTimeGetSeconds(position)
        self.dictionary["mediaViewStartingPosition"] = String(self.mediaViewStartingPosition)
        
        AwsS3.shared.uploadAnalytics(sessionId: self.sessionId,
                                     timestamp: self.dictionary["timeStarted"]! + "-B",
                                     prefix: "VideoBegV1",
                                     dictionary: self.dictionary, complete: { (error) in
                                        if let err = error {
                                            print("Error in upload analytics \(err)")
                                        }
        })
    }
    
    func playEnd(position: CMTime, completed: Bool) {
        print("INSIDE PLAY END \(position)  \(completed)")
        self.dictionary.removeAll()
        self.dictionary["sessionId"] = self.sessionId
        self.dictionary["timeStarted"] = dateFormatter.string(from: self.timeStarted)
        let timeCompleted = Date()
        self.dictionary["timeCompleted"] = dateFormatter.string(from: timeCompleted)
        let duration: TimeInterval = timeCompleted.timeIntervalSince(self.timeStarted)
        self.dictionary["elapsedTime"] = String(round(duration * 1000) / 1000)
        let secondsPlay = CMTimeGetSeconds(position)
        let mediaTimeViewInSeconds = secondsPlay - self.mediaViewStartingPosition
        self.dictionary["mediaTimeViewInSeconds"] = String(mediaTimeViewInSeconds)
        self.dictionary["mediaViewCompleted"] = String(completed)
        
        AwsS3.shared.uploadAnalytics(sessionId: self.sessionId,
                                     timestamp: self.dictionary["timeCompleted"]! + "-E",
                                     prefix: "VideoEndV1",
                                     dictionary: self.dictionary,
                                     complete: { (error) in
                                        if let err = error {
                                          print("Error in upload of analytics \(err)")
                                        }
                                        
        })
    }
}

File: Plugins/VideoPlayer/src/android/app/src/main/java/com/shortsands/videoplayer/VideoAnalytics.java
package com.shortsands.videoplayer;

import android.content.Context;
import android.content.pm.PackageInfo;
import android.content.pm.PackageManager.NameNotFoundException;
import android.os.Build;
import android.util.Log;

import com.shortsands.aws.AwsS3Manager;
import com.shortsands.aws.UploadDataListener;

import org.json.JSONException;
import org.json.JSONObject;

import java.text.DateFormat;
import java.text.SimpleDateFormat;

import java.util.Date;
import java.util.Locale;
import java.util.TimeZone;
/**
*  VideoAnalytics.java
*  AnalyticsProto
*
*  Created by Gary Griswold on 6/29/17.
*  Copyright © 2017 ShortSands. All rights reserved.
*/
class VideoAnalytics {

    private static String TAG = "VideoAnalytics";
    private static DateFormat isoFormat = new SimpleDateFormat("yyyy-MM-dd'T'HH:mm:ss'Z'"); // Quoted "Z" to indicate UTC, no timezone offset

    private Context context;
    private String mediaSource;
    private String mediaId;
    private String languageId;
    private String silLang;
    private String sessionId;
    
    // Pass following from play to playEnd
    private Date timeStarted;
    private long mediaViewStartingPosition;

    VideoAnalytics(Context context,
                   String mediaSource,
                   String mediaId,
                   String languageId,
                   String silLang) {
        this.context = context;
        this.mediaSource = mediaSource;
        this.mediaId = mediaId;
        this.languageId = languageId;
        this.silLang = silLang;
        
        AnalyticsSessionId analyticsSessionId = new AnalyticsSessionId(context);
        this.sessionId = analyticsSessionId.getSessionId();

        this.mediaViewStartingPosition = 0L;

        TimeZone tz = TimeZone.getTimeZone("UTC");
        //TimeZone tz = TimeZone.getDefault();
        isoFormat.setTimeZone(tz);
    }
    
    void playStarted(long position) {
        JSONObject dictionary = new JSONObject();
        try {
            dictionary.put("sessionId", this.sessionId);
            dictionary.put("mediaSource", this.mediaSource);
            dictionary.put("mediaId", this.mediaId);
            dictionary.put("languageId", this.languageId);
            dictionary.put("silLang", this.silLang);

            Locale locale = Locale.getDefault();
            dictionary.put("language", locale.getLanguage());
            dictionary.put("country", locale.getCountry());
            dictionary.put("locale", locale.toString());

            dictionary.put("deviceType", "mobile");
            dictionary.put("deviceFamily", Build.MANUFACTURER);
            dictionary.put("deviceName", Build.MODEL);
            dictionary.put("deviceOS", "android");
            dictionary.put("osVersion", Build.VERSION.RELEASE);
            try {
                PackageInfo pInfo = this.context.getPackageManager().getPackageInfo(this.context.getPackageName(), 0);
                dictionary.put("appVersion", pInfo.versionName);
                dictionary.put("appName", pInfo.packageName);
            } catch(NameNotFoundException nnfe) {
                dictionary.put("appVersion", nnfe.toString());
                dictionary.put("appName", "");
            }
            this.timeStarted = new Date();
            String timeStartedStr = isoFormat.format(this.timeStarted);
            dictionary.put("timeStarted", timeStartedStr);

            dictionary.put("isStreaming", "true");
            this.mediaViewStartingPosition = position;
            dictionary.put("mediaViewStartingPosition", Double.toString(this.mediaViewStartingPosition / 1000.0));

            Log.d(TAG, "BEGIN " + dictionary.toString());

            UploadDataListener listener = new UploadDataListener();
            AwsS3Manager.findSS().uploadAnalytics(this.sessionId, timeStartedStr + "-B", "VideoBegV1", dictionary.toString(2), listener);
        } catch(JSONException ex) {
            Log.e(TAG, "Error building Analytics Begin " + ex.toString());
        }
    }
    
    void playEnded(long position, boolean completed) {
        Log.d(TAG, "INSIDE PLAY END ");
        JSONObject dictionary = new JSONObject();
        try {
            dictionary.put("sessionId", this.sessionId);
            dictionary.put("timeStarted", isoFormat.format(this.timeStarted));
            Date timeCompleted = new Date();
            String timeCompletedStr = isoFormat.format(timeCompleted);
            dictionary.put("timeCompleted", timeCompletedStr);
            long duration = (timeCompleted.getTime() - this.timeStarted.getTime());
            dictionary.put("elapsedTime", Double.toString(duration / 1000.0));
            long mediaTimeViewInSeconds = position - this.mediaViewStartingPosition;
            dictionary.put("mediaTimeViewInSeconds", Double.toString(mediaTimeViewInSeconds / 1000.0));
            dictionary.put("mediaViewCompleted", Boolean.toString(completed));

            Log.d(TAG, "END " + dictionary.toString());

            UploadDataListener listener = new UploadDataListener();
            AwsS3Manager.findSS().uploadAnalytics(this.sessionId, timeCompletedStr + "-E", "VideoEndV1", dictionary.toString(2), listener);
        } catch(JSONException ex) {
            Log.e(TAG, "Error building Analytics End " + ex.toString());
        }
    }
}
